runtime/auth/signing/aws-signing-common/common/src/aws/smithy/kotlin/runtime/auth/signing/awssigning/common/AwsSigningConfig.kt

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package aws.smithy.kotlin.runtime.auth.signing.awssigning.common
+
+import aws.smithy.kotlin.runtime.auth.credentials.awscredentials.CredentialsProvider
+import aws.smithy.kotlin.runtime.time.Instant
+import kotlin.time.Duration
+
+typealias ShouldSignHeaderPredicate = (String) -> Boolean
+
+enum class AwsSigningAlgorithm {
+    SIGV4,
+    SIGV4_ASYMMETRIC,
+}
+
+enum class AwsSignatureType {
+    HTTP_REQUEST_VIA_HEADERS,
+    HTTP_REQUEST_VIA_QUERY_PARAMS,
+    HTTP_REQUEST_CHUNK,
+    HTTP_REQUEST_EVENT,
+}
+
+sealed class BodyHashSource(open val hash: String?) {
```

Comment:
I didn't like the old `signedBodyValue` because it was a nullable `string` but not every valid `string` is a valid `signedBodyValue`. So I made it an ADT to explicitly capture the use cases we support for signing.

And yeah, naming things is hard. One of these subtypes indicates that the hash is not yet known and should be calculated from the source. The others indicate effectively constant values. Taxonomically, that indicated to me that this overall type was modeling a _source_ of hash information, which may be explicit or not. Would `BodyHash` be clearer?

Reply:
I don't mind the ADT, I agree it's more clear with that. It was more of a naming question. `BodyHash` may be slightly clearer but I'll leave it up to you